File: gol/Domain/Entities/Person.cs
namespace gol.Domain.Entities;

public class Person
{
    private DateTime _currentDate;

    public string FirstName { get; private set; }
    public string LastName { get; private set; }
    public string Name => $"{FirstName} {LastName}";
    public string DisplayName => HasPartner && Partner!.LastName != LastName
        ? $"{FirstName} {LastName} ({Partner.LastName})"
        : Name;
    public DateTime BirthDate { get; private set; }
    public string Location { get; private set; }
    public string Gender { get; private set; }
    public IReadOnlyList<string> Licenses { get; private set; }
    public string EducationLevel { get; private set; }
    public IReadOnlyList<string> Hobbies { get; private set; }
    public Person? Partner { get; private set; }
    public DateTime? RelationshipStartDate { get; private set; }
    public bool IsAlive { get; private set; }
    public DateTime? DateOfDeath { get; private set; }
    public int MaxChildren { get; private set; }
    private List<Person> _children = new List<Person>();
    public IReadOnlyList<Person> Children => _children.AsReadOnly();
    public int NumberOfChildren => _children.Count;
    public DateTime? LastChildBirthDate => _children.Count > 0 ? _children[_children.Count - 1].BirthDate : null;
    public Person? Parent1 { get; private set; }
    public Person? Parent2 { get; private set; }
    public bool HasParents => Parent1 != null && Parent2 != null;

    public IReadOnlyList<Person> Siblings
    {
        get
        {
            if (!HasParents) return new List<Person>().AsReadOnly();

            var siblings = new List<Person>();
            if (Parent1 != null)
            {
                siblings.AddRange(Parent1.Children.Where(c => c != this));
            }
            return siblings.Distinct().ToList().AsReadOnly();
        }
    }

    public int Age => _currentDate.Year - BirthDate.Year - (_currentDate.DayOfYear < BirthDate.DayOfYear ? 1 : 0);
    public bool HasPartner => Partner != null;
    public bool IsLookingForPartner => Age >= 18 && !HasPartner && IsAlive;
    public TimeSpan RelationshipDuration => HasPartner && RelationshipStartDate.HasValue
        ? _currentDate - RelationshipStartDate.Value
        : TimeSpan.Zero;

    public Person(string firstName, string lastName, DateTime birthDate, string location, string gender,
                  List<string> licenses, string educationLevel, List<string> hobbies)
    {
        FirstName = firstName ?? throw new ArgumentNullException(nameof(firstName));
        LastName = lastName ?? throw new ArgumentNullException(nameof(lastName));
        BirthDate = birthDate;
        Location = location ?? throw new ArgumentNullException(nameof(location));
        Gender = gender ?? throw new ArgumentNullException(nameof(gender));
        Licenses = licenses?.AsReadOnly() ?? new List<string>().AsReadOnly();
        EducationLevel = educationLevel ?? throw new ArgumentNullException(nameof(educationLevel));
        Hobbies = hobbies?.AsReadOnly() ?? new List<string>().AsReadOnly();
        _currentDate = birthDate; // Initialize to birth date, will be updated as simulation progresses
        IsAlive = true;
        DateOfDeath = null;
        MaxChildren = 0; // Will be set when entering a relationship
    }

    public void AdvanceDay(DateTime newDate)
    {
        _currentDate = newDate;
    }

    public void SetPartner(Person partner, DateTime relationshipStartDate)
    {
        if (partner == null)
            throw new ArgumentNullException(nameof(partner));

        if (HasPartner && Partner != partner)
            throw new InvalidOperationException($"{Name} already has a partner.");

        if (partner.HasPartner && partner.Partner != this)
            throw new InvalidOperationException($"{partner.Name} already has a partner.");

        Partner = partner;
        RelationshipStartDate = relationshipStartDate;
    }

    public void SetMaxChildren(int maxChildren)
    {
        MaxChildren = maxChildren;
    }

    public void AddChild(Person child)
    {
        _children.Add(child);
    }

    public void SetParents(Person parent1, Person parent2)
    {
        Parent1 = parent1;
        Parent2 = parent2;
    }

    public bool IsCompatibleWith(Person other, int maxAgeDifference = 4)
    {
        if (other == null) return false;
        if (!IsLookingForPartner || !other.IsLookingForPartner) return false;

        var ageDifference = Math.Abs(Age - other.Age);
        return ageDifference <= maxAgeDifference;
    }

    public void Die(DateTime dateOfDeath)
    {
        IsAlive = false;
        DateOfDeath = dateOfDeath;
    }
}

File: gol/Web/Person.js
class Person {
    constructor(firstName, lastName, birthDate, gender, hobbies) {
        // Personal Information
        this.firstName = firstName;
        this.lastName = lastName;
        this.gender = gender;
        this.hobbies = hobbies;

        // Lifecycle & State
        this.birthDate = new Date(birthDate);
        this.currentDate = new Date(birthDate);
        this.cachedAge = 0;
        this.isAlive = true;
        this.dateOfDeath = null;

        // Partnership
        this.partner = null;
        this.relationshipStartDate = null;

        // Children
        this.maxChildren = 0;
        this.children = [];
        this.lastChildBirthDate = null;

        // Parents & Siblings
        this.parent1 = null;
        this.parent2 = null;
        this.cachedSiblings = null;
    }

    get name() {
        return `${this.firstName} ${this.lastName}`;
    }

    get displayName() {
        if (this.hasPartner && this.partner.lastName !== this.lastName) {
            return `${this.firstName} ${this.lastName} (${this.partner.lastName})`;
        }
        return this.name;
    }

    get age() {
        return this.cachedAge;
    }

    get hasPartner() {
        return this.partner !== null;
    }

    get isLookingForPartner() {
        return this.age >= 18 && !this.hasPartner && this.isAlive;
    }

    get relationshipDuration() {
        if (this.hasPartner && this.relationshipStartDate) {
            return this.currentDate - this.relationshipStartDate;
        }
        return 0;
    }

    get numberOfChildren() {
        return this.children.length;
    }

    get hasParents() {
        return this.parent1 !== null && this.parent2 !== null;
    }

    get siblings() {
        if (this.cachedSiblings !== null) {
            return this.cachedSiblings;
        }

        if (!this.hasParents) {
            this.cachedSiblings = [];
            return this.cachedSiblings;
        }

        const siblings = [];
        if (this.parent1) {
            for (const child of this.parent1.children) {
                if (child !== this) {
                    siblings.push(child);
                }
            }
        }
        this.cachedSiblings = siblings;
        return this.cachedSiblings;
    }

    advanceDay(newDate) {
        this.currentDate = new Date(newDate);
        const age = this.currentDate.getFullYear() - this.birthDate.getFullYear();
        const dayOfYear = this.getDayOfYear(this.currentDate);
        const birthDayOfYear = this.getDayOfYear(this.birthDate);
        this.cachedAge = dayOfYear < birthDayOfYear ? age - 1 : age;
    }

    getDayOfYear(date) {
        const start = new Date(date.getFullYear(), 0, 0);
        const diff = date - start;
        const oneDay = 1000 * 60 * 60 * 24;
        return Math.floor(diff / oneDay);
    }

    setPartner(partner, relationshipStartDate) {
        if (!partner) {
            throw new Error('Partner cannot be null');
        }

        if (this.hasPartner && this.partner !== partner) {
            throw new Error(`${this.displayName} already has a partner.`);
        }

        if (partner.hasPartner && partner.partner !== this) {
            throw new Error(`${partner.displayName} already has a partner.`);
        }

        this.partner = partner;
        this.relationshipStartDate = new Date(relationshipStartDate);
    }

    setMaxChildren(maxChildren) {
        this.maxChildren = maxChildren;
    }

    addChild(child) {
        this.children.push(child);
        this.lastChildBirthDate = new Date(child.birthDate);
        this.cachedSiblings = null;
    }

    setParents(parent1, parent2) {
        this.parent1 = parent1;
        this.parent2 = parent2;
    }

    isCompatibleWith(other, maxAgeDifference = 4) {
        if (!other) {
            return false;
        }

        if (!this.isLookingForPartner || !other.isLookingForPartner) {
            return false;
        }

        const ageDifference = Math.abs(this.age - other.age);
        return ageDifference <= maxAgeDifference;
    }

    die(dateOfDeath) {
        this.isAlive = false;
        this.dateOfDeath = new Date(dateOfDeath);
    }
}
